src/Remote/Importer.php

```diff
@@ -0,0 +1,162 @@
+use Castor\Remote\Exception\InvalidImportFormat;
+use Psr\Log\LoggerInterface;
+
+use function Castor\import;
+
+/** @internal */
+class Importer
+{
+    public function __construct(
+        private readonly LoggerInterface $logger,
+        private readonly Composer $composer,
+        /** @var array<string, array<string, array<string|null>>> */
+        private array $imports = [],
+    ) {
+    }
+
+    /**
+     * @param ?array{
+     *     url?: string,
+     *     type?: "git" | "svn",
+     *     reference?: string,
+     * } $source
+     */
+    public function importFunctionsFrom(string $scheme, string $package, ?string $file = null, ?string $version = null, ?string $vcs = null, ?array $source = null): void
+    {
+        if (!$this->allowsRemote()) {
+            throw new ImportError(sprintf('Remote imports are disabled, skipping import of "%s".', $package));
+        }
+
+        if (isset($this->imports[$package]) && ($requiredVersion = array_key_first($this->imports[$package])) !== $version) {
+            throw new ImportError(sprintf('The package "%s" is already required in version "%s", could not require it in version "%s"', $package, $requiredVersion, $version));
+        }
+
+        if (!preg_match('#^(?<organization>[^/]+)/(?<repository>[^/]+)$#', $package)) {
+            throw new InvalidImportFormat(sprintf('The import path must be formatted like this: "%s://<organization>/<repository>".', $scheme));
+        }
+
+        if ('composer' === $scheme) {
+            if (null !== $source) {
+                throw new InvalidImportFormat('The "source" argument is not supported for Composer/Packagist packages.');
+            }
+
+            $this->importPackageWithComposer($package, version: $version ?? '*', repositoryUrl: $vcs, file: $file);
+
+            return;
+        }
+
+        if ('package' === $scheme) {
+            if (null !== $version || null !== $vcs) {
+                throw new InvalidImportFormat('The "source" and "vcs" arguments are not supported for non-Composer packages.');
+            }
+            if (null === $source) {
+                throw new InvalidImportFormat('The "source" argument is required for non-Composer packages.');
+            }
+
+            $this->importPackageWithComposer($package, version: 'v1', source: $source, file: $file);
+
+            return;
+        }
+
+        throw new InvalidImportFormat(sprintf('The import scheme "%s" is not supported.', $scheme));
+    }
+
+    public function fetchPackages(): void
+    {
+        if (!$this->imports) {
+            $this->composer->remove();
+
+            return;
+        }
+
+        $input = GlobalHelper::getApplication()->getInput();
+
+        // Need to look for the raw options as the input is not yet parsed
+        $forceUpdate = null === $input->getParameterOption('--update-remotes', true);
+        $displayProgress = null !== $input->getParameterOption('--no-interaction', true);
```

Comment:
Something is strange here

```console
$ >…egoire/dev/github.com/jolicode/castor(import-with-composer) time castor --update-remotes > /dev/null

real    0m1,971s
user    0m0,357s
sys     0m0,061s
$ >…egoire/dev/github.com/jolicode/castor(import-with-composer) time castor --update-remotes -n > /dev/null

real    0m0,098s
user    0m0,059s
sys     0m0,037s
```

with -n, it's instant.

Reply:
The condition was wrong I think, I must have check against true, not null, as explained in another comment. But I changed the logic to only hide the display on the list command (to not break tools parsing the output (like the bin/generate-tests.php f.e.)